### PR TITLE
Speed up helm-ag--edit-commit using with-temp-buffer

### PR DESCRIPTION
**Pull request by Namburgesas**
Committing changes after editing search results in Helm can be very slow because the major-mode for each modified file is loaded, causing things like LSP to be loaded too. (The code currently uses find-file-noselect.) Using with-temp-buffer is the standard way to avoid loading anything when editing files non-interactively.

### CODE REVIEW

**Comment by Namburgesas** [comment]
@jcs090218 @syohex any thoughts on this? Small change for a huge boost in performance (minutes -> <1s for a large edit).

**Comment by syohex** [comment]
- I suppose this change makes losing file content if user uses helm-ag edit feature with some unsaved buffers. Because `insert-file-contents` returns saved file content, not current buffer content. However I'm not sure there is such case.
- Please fix byte-compile warnings
- I no longer maintain any emacs packages. @jcs090218  please decide to accept  or decline this PR

**Comment by syohex** [comment]
BTW is it enough to use `rawfile` option of `find-file-noselect` against un-opened files for avoiding loading heavy major/minor mode. This PR causes writing file multiple times if multiple lines in file are edited.

**Comment by jcs090218** [comment]
Since I am not the author of the package yet I am the maintainer of the package. I will accept PR community wise. 😃 

Here is more comments from @syohex.
> * I suppose this change makes losing file content if user uses helm-ag edit feature with some unsaved buffers. Because insert-file-contents returns saved file content, not current buffer content. However I'm not sure there is such case.

I am not sure about this but I would say that I don't have big issue with this. 😕 

> * Please fix byte-compile warnings. I suppose code looks good except byte-compile warnings?

Can you fix the byte-compile warnings? Thanks!

> BTW is it enough to use rawfile option of find-file-noselect against un-opened files for avoiding loading heavy major/minor mode. This PR causes writing file multiple times if multiple lines in file are edited.

I agreed with this.

**Comment by Namburgesas** [comment]
Sorry about the late reply. Thanks for the great feedback! 

>     * I suppose this change makes losing file content if user uses helm-ag edit feature with some unsaved buffers. Because `insert-file-contents` returns saved file content, not current buffer content. However I'm not sure there is such case.
Agree with @jcs090218 . I manually confirmed that the current implementation deals with saved file content, not current buffer content.
> 
>     * Please fix byte-compile warnings. I suppose code looks good except byte-compile warnings
Done.
> 
> BTW is it enough to use rawfile option of find-file-noselect against un-opened files for avoiding loading heavy major/minor mode. This PR causes writing file multiple times if multiple lines in file are edited.
I tested this, and it's a lot faster, but still orders of magnitude slower than using `with-temp-buffer`. Instead, I updated my implementation to batch edits per file.

File: helm-ag.el [hunk 1]
```diff
@@ -618,40 +618,44 @@ Default behaviour shows finish and result in mode-line."
...
+             (let ((deleted-lines (gethash curr-file line-deletes 0))
+                   (deleted (and ovs (overlay-get (car ovs) 'helm-ag-deleted))))
+               (forward-line (- line 1 deleted-lines))
+               (delete-region (line-beginning-position) (line-end-position))
+               (if (not deleted)
+                   (insert body)
+                 (let ((beg (point)))
+                   (forward-line 1)
+                   (delete-region beg (point))
+                   (puthash curr-file (1+ deleted-lines) line-deletes))))))
```
Comment (Namburgesas) [inline helm-ag.el:654]: This part is unchanged.

Comment (Namburgesas) [inline helm-ag.el:654]: This part is unchanged.

File: helm-ag.el [hunk 2]
```diff
@@ -618,40 +618,44 @@ Default behaviour shows finish and result in mode-line."
...
-      (dolist (buf saved-buffers)
-        (with-current-buffer buf
-          (save-buffer))))
-    (dolist (buf open-buffers)
-      (unless (memq buf kept-buffers)
-        (kill-buffer buf)))
+             (ovs (overlays-at (line-beginning-position)))
+             (lines-list (gethash file files-to-lines)))
+        (if (not (file-writable-p file))
+            (cl-incf read-only-files)
```
Comment (Namburgesas) [inline helm-ag.el:632]: Keeping track of read-only files as before.

File: helm-ag.el [hunk 3]
```diff
@@ -618,40 +618,44 @@ Default behaviour shows finish and result in mode-line."
...
-              (cl-pushnew (current-buffer) saved-buffers))))))
-    (when helm-ag-edit-save
-      (dolist (buf saved-buffers)
-        (with-current-buffer buf
-          (save-buffer))))
-    (dolist (buf open-buffers)
-      (unless (memq buf kept-buffers)
-        (kill-buffer buf)))
+             (ovs (overlays-at (line-beginning-position)))
+             (lines-list (gethash file files-to-lines)))
```
Comment (Namburgesas) [inline helm-ag.el:630]: Just added a variable to the let* statement to store a list of changed lines for each file.

**Comment by jcs090218** [comment]
> I tested this, and it's a lot faster, but still orders of magnitude slower than using with-temp-buffer. Instead, I updated my implementation to batch edits per file.

Okay, I think this is the thing I concerned the most. Once this is solved, I think this PR is good to merged. I wonder why the changes is so... broad? I think git somehow mess up with the line changes/tracking. Anyway, thanks for the patch! Merged!

**Pull request merged**